### ISSUE TITLE
ValueReader Handling is Broken

### ISSUE THREAD

**Issue by saden1** [modelmapper/modelmapper#188]
There seems to be a bug in how custom value readers are being handled. Here is a test case that demonstrates the issue.

`````` java
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import static org.assertj.core.api.Assertions.assertThat;
import org.junit.Before;
import org.junit.Test;
import org.modelmapper.ModelMapper;
import org.modelmapper.config.Configuration.AccessLevel;
import org.modelmapper.convention.MatchingStrategies;
import org.modelmapper.convention.NameTokenizers;
import org.modelmapper.convention.NamingConventions;
import org.modelmapper.spi.ValueReader;

public class ModelMapperTest {

    ModelMapper cut;

    @Before
    public void init() {
        cut = new ModelMapper();
        cut.getConfiguration()
                .addValueReader(new ItemValueReader())
                .setFieldMatchingEnabled(true)
                .setFieldAccessLevel(AccessLevel.PRIVATE)
                .setMatchingStrategy(MatchingStrategies.STRICT)
                .setSourceNamingConvention(NamingConventions.JAVABEANS_ACCESSOR)
                .setDestinationNamingConvention(NamingConventions.JAVABEANS_MUTATOR)
                .setSourceNameTokenizer(NameTokenizers.CAMEL_CASE)
                .setDestinationNameTokenizer(NameTokenizers.CAMEL_CASE);
    }

    @Test
    public void givenMapAsSourceMapShouldCopySourceToDestination() {
        Map<String, String> source = ImmutableMap.<String, String>builder()
                .put("username", "testuser")
                .put("test", "test")
                .build();

        User destination = cut.map(source, User.class);

        assertThat(destination.getUsername())
                .as("username is in the source map and should be present in the destination")
                .isEqualTo("testuser");

        assertThat(destination.getPassword())
                .as("password is not in the source map and should not be present in the destination")
                .isNull();

        //lets create a new map
        source = ImmutableMap.<String, String>builder()
                .put("username", "testuser")
                .put("password", "testpassword")
                .build();

        destination = cut.map(source, User.class);

        assertThat(destination.getUsername())
                .as("username is in the source item and should be present in the destination")
                .isEqualTo("testuser");

        assertThat(destination.getPassword())
                .as("password is in the source item and should be present in the destination")
                .isEqualTo("testpassword");

    }

    @Test
    public void givenItemAsSourceMapShouldCopySourceToDestination() {
        Item source = new Item()
                .put("username", "testuser");

        User destination = cut.map(source, User.class);

        assertThat(destination.getUsername())
                .as("username is in the source item and should be present in the destination")
                .isEqualTo("testuser");

        assertThat(destination.getPassword())
                .as("password is not in the source item and should not be present in the destination")
                .isNull();

        //lets create a new item
        source = new Item()
                .put("username", "testuser")
                .put("password", "testpassword");

        destination = cut.map(source, User.class);

        assertThat(destination.getUsername())
                .as("username is in the source item and should be present in the destination")
                .isEqualTo("testuser");

        assertThat(destination.getPassword())
                .as("password is in the source item and should be present in the destination")
                .isEqualTo("testpassword");

    }

    public class ItemValueReader implements ValueReader<Item> {

        List<String> columns = ImmutableList.of("username", "password");

        @Override
        public Object get(Item t, String string) {
            return t.get(string);
        }

        @Override
        public Collection<String> memberNames(Item t) {
            return columns;
        }

    }

    public static class Item {

        private final Map<String, Object> store = new HashMap<>();

        public Object get(String key) {
            return store.get(key);
        }

        public Item put(String key, Object value) {
            store.put(key, value);

            return this;
        }

        public Collection<String> getKeys() {
            return store.keySet();
        }
    }

    public static class User {

        private String username;
        private String password;

        public User() {
        }

        public String getUsername() {
            return username;
        }

        public void setUsername(String username) {
            this.username = username;
        }

        public String getPassword() {
            return password;
        }

        public void setPassword(String password) {
            this.password = password;
        }

    }

}```
``````

**Comment by carofe82** [comment]
Same problem here

**Comment by chhsiao90** [comment]
Fixed by #210
